hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/FileSourceCleaner.java

```diff
@@ -0,0 +1,187 @@
+      cleanerPool.get().submit(() -> cleanTask(file));
+    } else {
+      cleanTask(file);
+    }
+  }
+
+  abstract void cleanTask(String file);
+
+  private static class FileSourceRemover extends FileSourceCleaner {
+    private final FileSystem fs;
+    public FileSourceRemover(TypedProperties props, FileSystem fs) {
+      super(props);
+      this.fs = fs;
+    }
+
+    @Override
+    void cleanTask(String file) {
+      LOG.info(String.format("Removing %s...", file));
+      try {
+        if (fs.delete(new Path(file), false)) {
+          LOG.info(String.format("Successfully remove up %s", file));
+        } else {
+          LOG.warn(String.format("Failed to remove %s", file));
+        }
+      } catch (IOException e) {
+        LOG.error(String.format("Failed to remove %s", file), e);
+      }
+    }
+  }
+
+  private static class FileSourceArchiver extends FileSourceCleaner {
+    private final FileSystem fs;
+    private final Path archiveDir;
+    private final Path sourceRootDir;
+
+    public FileSourceArchiver(TypedProperties props, FileSystem fs) {
+      super(props);
+      this.fs = fs;
+      this.archiveDir = new Path(props.getString(Config.FILE_SOURCE_ARCHIVE_DIR_KEY));
+      this.sourceRootDir = new Path(props.getString(ROOT_INPUT_PATH_PROP));
+      ValidationUtils.checkArgument(!isSubDir(archiveDir, sourceRootDir),
+          String.format("%s must not be child of %s", Config.FILE_SOURCE_ARCHIVE_DIR_KEY, ROOT_INPUT_PATH_PROP));
+    }
+
+    private boolean isSubDir(Path childDir, Path parentDir) {
+      while (childDir != null) {
+        if (childDir.equals(parentDir)) {
+          return true;
+        }
+        childDir = childDir.getParent();
+      }
+      return false;
+    }
+
+    @Override
+    void cleanTask(String file) {
+      try {
+        final Path original = new Path(file);
+        final Path fileDir = original.getParent();
+        Path relativeDir = getRelativeDir(fileDir, sourceRootDir);
+        final Path newDir = new Path(archiveDir, relativeDir);
+        LOG.info("Creating directory if not existent: " + newDir.toString());
+        fs.mkdirs(newDir);
+
+        final Path newFile = new Path(newDir, original.getName());
+        LOG.info(String.format("Renaming: %s to %s", original.toString(), newFile));
+        if (fs.rename(original, newFile)) {
+          LOG.info(String.format("Successfully archive %s", file));
+        } else {
+          LOG.warn(String.format("Failed to archive %s", file));
+        }
+      } catch (IOException e) {
+        LOG.error(String.format("Failed to archive %s", file), e);
+      }
+    }
+
+    private Path getRelativeDir(Path childPath, Path parentPath) {
+      LinkedList<String> paths = new LinkedList<>();
+      while (childPath != null && !childPath.equals(parentPath)) {
+        paths.addFirst(childPath.getName());
+        childPath = childPath.getParent();
+      }
+      return new Path(paths.isEmpty() ? "." : String.join("/", paths));
```

Comment:
if file to be archived is in source root dir itself, this moves it to "archive_folder/./" right. May I know the rational ?

Reply:
The files are archived to the new location with relative to the source root directory. e.g. src/foo/x => archive/foo/x etc. So yes if a file is in the source root dir itself it will be moved to the archive root dir.

Reply:
yes, but trying to understand why move it to "archive_folder/./" folder rather than "archive_folder/". can you please help me understand.

Reply:
Right, actually the result will never be archive_folder/./ since `new Path(".")` returns `Path("")`. I can replace "." with "" if it causes confusion. Initially I thought "." would be more obvious that we return the current parent directory without any sub-dir.

Reply:
Hmmm turned out you can't initialize a Path with "". Would `new Path(paths.isEmpty() ? "." : String.join("/", paths))` be ok still?

Reply:
thanks for clarifying. actually we can leave it as is.